File: android/pi-tools/src/main/kotlin/com/pimobile/tools/EditFileTool.kt
package com.pimobile.tools

import com.pimobile.agent.*
import kotlinx.serialization.json.*
import java.io.File

class EditFileTool(private val sandboxDir: File) : Tool {

    override val name = "edit_file"
    override val description = "Apply search/replace edits to a file. Each edit specifies an exact string to find and its replacement."
    override val parametersSchema = buildJsonObject {
        put("type", "object")
        putJsonObject("properties") {
            putJsonObject("path") {
                put("type", "string")
                put("description", "File path relative to sandbox directory")
            }
            putJsonObject("edits") {
                put("type", "array")
                putJsonObject("items") {
                    put("type", "object")
                    putJsonObject("properties") {
                        putJsonObject("search") {
                            put("type", "string")
                            put("description", "Exact text to search for")
                        }
                        putJsonObject("replace") {
                            put("type", "string")
                            put("description", "Text to replace with")
                        }
                    }
                    putJsonArray("required") { add("search"); add("replace") }
                }
            }
        }
        putJsonArray("required") { add("path"); add("edits") }
    }

    override suspend fun execute(input: JsonObject): AgentToolResult {
        val path = input["path"]?.jsonPrimitive?.contentOrNull
            ?: return AgentToolResult("", "Error: 'path' is required", isError = true)
        val editsArray = input["edits"]?.jsonArray
            ?: return AgentToolResult("", "Error: 'edits' is required", isError = true)

        return try {
            val file = File(sandboxDir, path).canonicalFile
            if (!file.canonicalPath.startsWith(sandboxDir.canonicalPath)) {
                return AgentToolResult("", "Error: Path is outside sandbox", isError = true)
            }
            if (!file.exists()) {
                return AgentToolResult("", "Error: File not found: $path", isError = true)
            }

            val originalContent = file.readText()
            var currentContent = originalContent
            val appliedEdits = mutableListOf<String>()

            for (editJson in editsArray) {
                val editObj = editJson.jsonObject
                val search = editObj["search"]?.jsonPrimitive?.contentOrNull ?: continue
                val replace = editObj["replace"]?.jsonPrimitive?.contentOrNull ?: continue

                val occurrences = countOccurrences(currentContent, search)
                when {
                    occurrences == 0 -> {
                        return AgentToolResult(
                            "",
                            "Error: Search string not found in file:\n$search",
                            isError = true
                        )
                    }
                    occurrences > 1 -> {
                        return AgentToolResult(
                            "",
                            "Error: Search string found $occurrences times (must be unique):\n$search",
                            isError = true
                        )
                    }
                }

                currentContent = currentContent.replaceFirst(search, replace)
                appliedEdits.add("- Replaced ${search.length} chars with ${replace.length} chars")
            }

            file.writeText(currentContent)

            val hunks = generateDiffHunks(originalContent, currentContent)
            val diffText = formatUnifiedDiff(path, hunks)

            AgentToolResult(
                "",
                "Applied ${appliedEdits.size} edit(s) to $path\n$diffText",
                ToolResultDetails.Diff(path, hunks)
            )
        } catch (e: Exception) {
            AgentToolResult("", "Error editing file: ${e.message}", isError = true)
        }
    }

    private fun countOccurrences(text: String, search: String): Int {
        var count = 0
        var index = 0
        while (true) {
            index = text.indexOf(search, index)
            if (index == -1) break
            count++
            index += search.length
        }
        return count
    }

    private fun generateDiffHunks(original: String, modified: String): List<DiffHunk> {
        val oldLines = original.lines()
        val newLines = modified.lines()
        val hunks = mutableListOf<DiffHunk>()

        var i = 0
        var j = 0
        while (i < oldLines.size || j < newLines.size) {
            if (i < oldLines.size && j < newLines.size && oldLines[i] == newLines[j]) {
                i++
                j++
                continue
            }

            val hunkStartOld = i
            val hunkStartNew = j
            val lines = mutableListOf<DiffLine>()

            val contextStart = maxOf(0, i - 3)
            for (c in contextStart until i) {
                lines.add(DiffLine(DiffLineType.CONTEXT, oldLines[c]))
            }

            while (i < oldLines.size && (j >= newLines.size || oldLines[i] != newLines.getOrNull(j))) {
                lines.add(DiffLine(DiffLineType.REMOVE, oldLines[i]))
                i++
            }
            while (j < newLines.size && (i >= oldLines.size || newLines[j] != oldLines.getOrNull(i))) {
                lines.add(DiffLine(DiffLineType.ADD, newLines[j]))
                j++
            }

            val contextEnd = minOf(oldLines.size, i + 3)
            for (c in i until contextEnd) {
                if (c < oldLines.size) {
                    lines.add(DiffLine(DiffLineType.CONTEXT, oldLines[c]))
                }
            }

            val removedCount = lines.count { it.type == DiffLineType.REMOVE } +
                    lines.count { it.type == DiffLineType.CONTEXT }
            val addedCount = lines.count { it.type == DiffLineType.ADD } +
                    lines.count { it.type == DiffLineType.CONTEXT }

            hunks.add(DiffHunk(
                startLineOld = hunkStartOld + 1,
                countOld = removedCount,
                startLineNew = hunkStartNew + 1,
                countNew = addedCount,
                lines = lines
            ))

            i = contextEnd
            j += (contextEnd - hunkStartOld) - lines.count { it.type == DiffLineType.REMOVE } +
                    lines.count { it.type == DiffLineType.ADD }
            break
        }

        return hunks
    }

    private fun formatUnifiedDiff(path: String, hunks: List<DiffHunk>): String {
        val sb = StringBuilder()
        sb.appendLine("--- a/$path")
        sb.appendLine("+++ b/$path")
        for (hunk in hunks) {
            sb.appendLine("@@ -${hunk.startLineOld},${hunk.countOld} +${hunk.startLineNew},${hunk.countNew} @@")
            for (line in hunk.lines) {
                val prefix = when (line.type) {
                    DiffLineType.CONTEXT -> " "
                    DiffLineType.ADD -> "+"
                    DiffLineType.REMOVE -> "-"
                }
                sb.appendLine("$prefix${line.content}")
            }
        }
        return sb.toString()
    }
}

File: ios/PiTools/Sources/PiTools/EditFileTool.swift
import Foundation
import PiAgentCore
import PiAI

public struct EditFileTool: Tool, Sendable {
    public let name = "edit_file"
    public let description = "Edit a file by performing exact search and replace. The old_string must match exactly one location in the file."

    private let sandboxURL: URL

    public init(sandboxURL: URL) {
        self.sandboxURL = sandboxURL
    }

    public var parametersSchema: JSONValue {
        .object([
            "type": .string("object"),
            "required": .array([.string("path"), .string("old_string"), .string("new_string")]),
            "properties": .object([
                "path": .object([
                    "type": .string("string"),
                    "description": .string("Path to the file to edit, relative to sandbox"),
                ]),
                "old_string": .object([
                    "type": .string("string"),
                    "description": .string("The exact string to search for"),
                ]),
                "new_string": .object([
                    "type": .string("string"),
                    "description": .string("The replacement string"),
                ]),
            ]),
        ])
    }

    public func execute(input: JSONValue) async throws -> AgentToolResult {
        guard let path = input["path"]?.stringValue else {
            return AgentToolResult(output: "Error: 'path' parameter is required", isError: true)
        }
        guard let oldString = input["old_string"]?.stringValue else {
            return AgentToolResult(output: "Error: 'old_string' parameter is required", isError: true)
        }
        guard let newString = input["new_string"]?.stringValue else {
            return AgentToolResult(output: "Error: 'new_string' parameter is required", isError: true)
        }

        let fileURL = sandboxURL.appendingPathComponent(path)

        guard fileURL.standardizedFileURL.path.hasPrefix(sandboxURL.standardizedFileURL.path) else {
            return AgentToolResult(output: "Error: Path is outside sandbox", isError: true)
        }

        let fm = FileManager.default
        guard fm.fileExists(atPath: fileURL.path) else {
            return AgentToolResult(output: "Error: File not found: \(path)", isError: true)
        }

        guard let data = fm.contents(atPath: fileURL.path),
              let content = String(data: data, encoding: .utf8)
        else {
            return AgentToolResult(output: "Error: Could not read file as text", isError: true)
        }

        // Count occurrences
        let occurrences = content.components(separatedBy: oldString).count - 1
        if occurrences == 0 {
            return AgentToolResult(output: "Error: old_string not found in file", isError: true)
        }
        if occurrences > 1 {
            return AgentToolResult(output: "Error: old_string found \(occurrences) times. Must match exactly once.", isError: true)
        }

        // Perform replacement
        let newContent = content.replacingOccurrences(of: oldString, with: newString)

        guard let newData = newContent.data(using: .utf8) else {
            return AgentToolResult(output: "Error: Could not encode result as UTF-8", isError: true)
        }

        do {
            try newData.write(to: fileURL)
        } catch {
            return AgentToolResult(output: "Error writing file: \(error.localizedDescription)", isError: true)
        }

        // Generate diff info
        let hunks = generateDiff(old: oldString, new: newString, in: content)

        return AgentToolResult(
            output: "Edited \(path): replaced 1 occurrence",
            details: .diff(path: path, hunks: hunks)
        )
    }

    private func generateDiff(old: String, new: String, in content: String) -> [DiffHunk] {
        let contentLines = content.components(separatedBy: "\n")
        let oldLines = old.components(separatedBy: "\n")
        let newLines = new.components(separatedBy: "\n")

        // Find the start line of the match
        let contentJoined = content
        guard let matchRange = contentJoined.range(of: old) else { return [] }
        let startLine = contentJoined[contentJoined.startIndex..<matchRange.lowerBound].components(separatedBy: "\n").count

        var diffLines: [DiffLine] = []
        for line in oldLines {
            diffLines.append(DiffLine(type: .remove, content: line))
        }
        for line in newLines {
            diffLines.append(DiffLine(type: .add, content: line))
        }

        return [DiffHunk(
            startLineOld: startLine,
            countOld: oldLines.count,
            startLineNew: startLine,
            countNew: newLines.count,
            lines: diffLines
        )]
    }
}
